builtin/providers/aws/resource_aws_s3_bucket_object.go

```diff
@@ -0,0 +1,112 @@
+package aws
+
+import (
+	"fmt"
+	"log"
+	"os"
+
+	"github.com/hashicorp/terraform/helper/schema"
+
+	"github.com/awslabs/aws-sdk-go/aws"
+	"github.com/awslabs/aws-sdk-go/aws/awsutil"
+	"github.com/awslabs/aws-sdk-go/service/s3"
+)
+
+func resourceAwsS3BucketObject() *schema.Resource {
+	return &schema.Resource{
+		Create: resourceAwsS3BucketObjectPut,
+		Read:   resourceAwsS3BucketObjectRead,
+		Update: resourceAwsS3BucketObjectPut,
+		Delete: resourceAwsS3BucketObjectDelete,
+
+		Schema: map[string]*schema.Schema{
+			"bucket": &schema.Schema{
+				Type:     schema.TypeString,
+				Required: true,
+				ForceNew: true,
+			},
+
+			"key": &schema.Schema{
+				Type:     schema.TypeString,
+				Required: true,
+				ForceNew: true,
+			},
+
+			"source": &schema.Schema{
+				Type:     schema.TypeString,
+				Required: true,
+				ForceNew: true,
+			},
+		},
+	}
+}
+
+func resourceAwsS3BucketObjectPut(d *schema.ResourceData, meta interface{}) error {
+	s3conn := meta.(*AWSClient).s3conn
+
+	bucket := d.Get("bucket").(string)
+	key := d.Get("key").(string)
+	source := d.Get("source").(string)
+
+	file, err := os.Open(source)
+
+	if err != nil {
+		d.SetId("")
+		return fmt.Errorf("Error opening S3 bucket object source(%s): %s", source, err)
+	}
+
+	resp, err := s3conn.PutObject(
+		&s3.PutObjectInput{
+			Bucket: aws.String(bucket),
+			Key:    aws.String(key),
+			Body:   file,
+		})
+
+	if err != nil {
+		d.SetId("")
+		return fmt.Errorf("Error putting object in S3 bucket (%s): %s", bucket, err)
+	}
+
+	d.SetId(*resp.ETag)
```

Comment:
I'm hesitant on using the `ETag` here, simply because the API mentions that if a user supplies their own key to encrypt the file(s), the `ETag` will *not* be the `md5`... it doesn't clearly say what it _will_ be in that case though. I worry that down the road we expand on this and then get into a situation where we're losing objects :/ 

Can we simply use `key` here, instead?